mso/resource_mso_schema_template_contract_filter.go

```diff
@@ -72,753 +75,245 @@ func resourceMSOTemplateContractFilter() *schema.Resource {
-										d.Set("directives", filterCont.S("directives").Data().([]interface{}))
-										found = true
-										break
-									}
-								}
-							}
-						}
-					} else if stateFilterType == "consumer_to_provider" {
-						if contractCont.Exists("filterRelationshipsConsumerToProvider") {
-							filtercount, _ := contractCont.ArrayCount("filterRelationshipsConsumerToProvider")
-							for k := 0; k < filtercount; k++ {
-								filterCont, err := contractCont.ArrayElement(k, "filterRelationshipsConsumerToProvider")
-								if err != nil {
-									return nil, fmt.Errorf("Unable to parse the filter Relationships Consumer To Provider list")
-								}
+func createMSOTemplateContractFilterPath(templateName, contractName, filterRelationshipType, name string) string {
+	return fmt.Sprintf("/templates/%s/contracts/%s/%s/%s", templateName, contractName, filterRelationshipType, name)
+}
 
-								if filterCont.Exists("filterRef") {
-									filRef := filterCont.S("filterRef").Data()
-									split := strings.Split(filRef.(string), "/")
-									if split[6] == stateName && split[4] == stateFilterTemplate && split[2] == stateFilterSchema {
-										d.Set("filter_name", split[6])
-										d.Set("filter_template_name", split[4])
-										d.Set("filter_schema_id", split[2])
-										d.Set("directives", filterCont.S("directives").Data().([]interface{}))
-										found = true
-										break
-									}
+func setContractFilterFromSchema(d *schema.ResourceData, schemaCont *container.Container, templateName, contractName, filterType, filterSchemaId, filterTemplateName, filterName string) error {
+	log.Printf("[DEBUG] %s: Beginning set contract filter", d.Id())
+	templates := schemaCont.Search("templates").Data()
+	if templates == nil || len(templates.([]interface{})) == 0 {
+		return fmt.Errorf("no templates found")
+	}
+
+	filterRelationshipType := getFilterRelationshipTypeMap()[filterType]
+
+	for _, template := range templates.([]interface{}) {
+		templateDetails := template.(map[string]interface{})
+		if templateDetails["name"].(string) == templateName {
+			for _, contract := range templateDetails["contracts"].([]interface{}) {
+				contractDetails := contract.(map[string]interface{})
+				if contractDetails["name"].(string) == contractName {
+					if filterRelationships, ok := contractDetails[filterRelationshipType]; filterRelationships != nil && ok {
+						for _, filterRelationship := range filterRelationships.([]interface{}) {
+							filterRelationshipMap := filterRelationship.(map[string]interface{})
+							if val, ok := filterRelationshipMap["filterRef"]; ok {
+								re := regexp.MustCompile("/schemas/(.*)/templates/(.*)/filters/(.*)")
+								match := re.FindStringSubmatch(val.(string))
+								if match[1] == filterSchemaId && match[2] == filterTemplateName && match[3] == filterName {
+									d.SetId(createMSOTemplateContractFilterPath(templateName, contractName, filterRelationshipType, filterName))
```

Comment:
fixed

Reply:
Ok the reason for setting it to path is re-use. The import statement has to deviate from that in order retrieve the correct information. Does this have to match the import id?

Reply:
@lhercot please share your thoughts.

Reply:
Yes, when using import you are providing the Id of the resources you are importing. This should be the same Id as the one set anywhere in the resource so you can use the current state value to remove and re-import.

Why are we changing that?

Reply:
@lhercot I changed this in the this import statement because my understanding was that we need to set the id to the path ( id ) which can be directly used for deletion. Would need to change logic of the parent contract resource then also.

A few point we would need to discuss:

1. How do we handle the referenced objects? In this case the filter which has a {filter_schema_id}/{filter_template_name}, do we keep this in the id/import statement? If so, do we need some additional function to extract the path from the id values?
2. How do we handle containers in the ndo schema that are not exposed to the user through UI but are needed to differentiate what the user needs to input. In this case the filter_type, which is inputted as bothWay, provider_to_consumer, consumer_to_provider but gets translated to filterRelationships, filterRelationshipsProviderToConsumer, filterRelationshipsConsumerToProvider.

Reply:
I have adjusted the id setting to be same as before, this would mean that id can not be used directly for destroy operation.